Make bison and flex required, move output to build dir, and generate on Windows (MSVC)

Simran-B opened the pull request:
### Scope & Purpose

Remove generated tokens and grammar files from repo, generate them in the build dir, and always generate them (not in maintainer mode only).

- [ ] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [x] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [ ] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [ ] Backports
  - [ ] Backport for 3.10: *(Please link PR)*
  - [ ] Backport for 3.9: *(Please link PR)*
  - [ ] Backport for 3.8: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

Simran-B commented on arangod/Aql/CMakeLists.txt:
```diff
@@ -194,30 +194,35 @@ target_compile_options(arango_aql PRIVATE
     VERBATIM
   )
 
+  # Find Git Bash on Windows. GITBASH_EXECUTABLE is not set for other OSes
```
Works fine under Linux in my test:
```
[ 70%] Generating grammar.cpp
/work/ArangoDB/arangod/Aql/grammar.y:2.1-18: warning: deprecated directive: ‘%name-prefix "Aql"’, use ‘%define api.prefix {Aql}’ [-Wdeprecated]
    2 | %name-prefix "Aql"
      | ^~~~~~~~~~~~~~~~~~
      | %define api.prefix {Aql}
/work/ArangoDB/arangod/Aql/grammar.y: warning: fix-its can be applied.  Rerun with option '--update'. [-Wother]
Consolidate compiler generated dependencies of target arango_aql
[ 70%] Building CXX object arangod/Aql/CMakeFiles/arango_aql.dir/grammar.cpp.o
[ 70%] Linking CXX static library ../../bin/libarango_aql.a
[ 77%] Built target arango_aql
```
It seems that an undefined `${GIBASH_EXECUTABLE}` is simply ignored as intended. If CMake turned this into a call `"" path/utils/bison-c.sh` it would fail, but it seems to fine 👍

Simran-B commented on arangod/Aql/CMakeLists.txt:
```diff
@@ -194,30 +194,35 @@ target_compile_options(arango_aql PRIVATE
-if (USE_MAINTAINER_MODE AND NOT MSVC)
-  set_directory_properties(PROPERTIES CLEAN_NO_CUSTOM "On")
+if (USE_MAINTAINER_MODE)
+  set_directory_properties(PROPERTIES CLEAN_NO_CUSTOM "On") # TODO: Remove this?
```
I'm not sure if this should be kept as-is or if it needs adjustments.

Simran-B commented on arangod/Aql/CMakeLists.txt:
```diff
@@ -194,30 +194,35 @@ target_compile_options(arango_aql PRIVATE
 
 if (CMAKE_COMPILER_IS_GNUCC)
-  set_source_files_properties(tokens.cpp PROPERTIES COMPILE_FLAGS -Wno-sign-compare)
+  set_source_files_properties(tokens.cpp PROPERTIES COMPILE_FLAGS -Wno-sign-compare) # TODO: Set directory?
```
I'm not sure if this should be adjusted/removed with the file now being generated in the build instead of the source dir.

Simran-B replied:
The function name might be misleading. I can't find a `binary` / `target` variant of the same command, so `source` might actually refer to source **files** and not the source **folder**. I'm not sure if it actually works, we might need to set the build dir arangod/Aql/ here, but who knows how CMake works.

mchacki reviewed (changes requested):
This change works for me.

However I would like to get the original checked in files:

	deleted:    arangod/Aql/grammar.cpp
	deleted:    arangod/Aql/tokens.cpp
	
Deleted by this PR. They should not be used anymore.
I tested this locally with above mentioned files erased and it compiled for me.
	
Thank you very much for this!

Simran-B commented:
`find_package()` wasn't called unconditionally, so that the variables for the Flex and Bison executables weren't set. Fixed.
Also fixed the include from `Aql/grammar.hpp` to `grammar.hpp`.

There are Jenkins errors on macOS and Windows:

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/bison: extra operand `/Users/jenkins/hw37-mac/oskar/work/ArangoDB/arangod/Aql/grammar.y'
Try `/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/bison --help' for more information.
make[2]: *** [arangod/Aql/grammar.cpp] Error 1
```

```
  bison: extra operand 'C:/gce-win-m7ivsv/oskar/work/ArangoDB/arangod/Aql/grammar.y'
  Try `bison --help' for more information.
```

This might be due to old binaries being used. macOS:

```
-- Found FLEX: /usr/bin/flex (found version "2.6.4") 
-- Found BISON: /usr/bin/bison (found version "2.3") 
```

Windows:
```
-- Found FLEX: C:/tools/DevKit2/bin/flex.exe (found version "2.6.3") 
-- Found BISON: C:/tools/DevKit2/bin/bison.exe (found version "2.7") 
```

The latter matches the choco package `winflexbison`: https://community.chocolatey.org/packages/winflexbison
There is a newer `winflexbison3` package (bison 3.7.4, flex 2.6.4): https://community.chocolatey.org/packages/winflexbison3
Latest Bison is 3.8.2 and Flex 2.6.4 is apparently the latest version since 2017.

Simran-B commented:
The invocation of Bison < 3 did not set an options variable, but because of quote marks, `''` remained in the call, leading to the extra operand error. Even with this fixed, there is another issue with Bison 2.x:
`error: %define variable 'parse.error' is not used`
It's not supported but can be substituted with `%error-verbose` as a workaround.

Simran-B commented:
Looks like my workaround for macOS and Windows doesn't work, or there's another issue with the specific versions of Bison:

```
grammar.y~:1.9-16: syntax error, unexpected identifier, expecting string
make[2]: *** [arangod/Aql/grammar.cpp] Error 1
```

@KVS85 Can we please upgrade Jenkins to use a Bison 3.x version?

Simran-B commented:
We dropped Windows and macOS support so this could be merged now but no one seems to be interested in making the lexer and parser generation build time artifacts.
